linfa-linear/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+//! # Linear Models
+//!
+//! `linfa-linear` aims to provide pure Rust implementations of
+//! popular linear regression algorithms.
+//!
+//! ## The Big Picture
+//!
+//! `linfa-linear` is a crate in the [`linfa`](https://crates.io/crates/linfa)
+//! ecosystem, a wider effort to bootstrap a toolkit for classical
+//! Machine Learning implemented in pure Rust, kin in spirit to
+//! Python's `scikit-learn`.
+//!
+//! ## Current state
+//!
+//! Right now `linfa-linear` provides ordinary least squares regression.
+//!
+//! ## Examples
+//!
+//! There is an usage example in the `examples/diabetes.rs` file, to run it
+//! run
+//!
+//! ```bash
+//! $ cargo run --features openblas --examples diabetes
+//! ```
+
+#![allow(non_snake_case)]
+use ndarray::{Array1, Array2, ArrayBase, Axis, Data, Ix1, Ix2, ScalarOperand};
+use ndarray_linalg::{Lapack, Scalar, Solve};
+use ndarray_stats::SummaryStatisticsExt;
+use num_traits::float::Float;
+
+/// An ordinary least squares linear regression model.
+///
+/// LinearRegression fits a linear model to minimize the residual sum of
+/// squares between the observed targets in the dataset, and the targets
+/// predicted by the linear approximation.
+///
+/// Ordinary least squares regression solves the overconstrainted model
+///
+/// y = Ax + b
+///
+/// by finding x and b which minimize the L_2 norm ||y - Ax - b||_2.
+///
+/// It currently uses the [Moore-Penrose pseudo-inverse]()
+/// to solve y - b = Ax.
+pub struct LinearRegression {
+    options: Options,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum Options {
+    None,
+    FitIntercept,
+    FitInterceptAndNormalize,
+}
+
+fn fit_intercept(options: Options) -> bool {
+    options == Options::FitIntercept || options == Options::FitInterceptAndNormalize
+}
+
+fn normalize(options: Options) -> bool {
+    options == Options::FitInterceptAndNormalize
+}
+
+/// A fitted linear regression model which can be used for making predictions.
+pub struct FittedLinearRegression<A> {
+    intercept: A,
+    params: Array1<A>,
+}
+
+/// Configure and fit a linear regression model
+impl LinearRegression {
+    /// Create a default linear regression model. By default, no intercept
+    /// will be fitted and the feature matrix will not be normalized.
+    /// To change this, call `fit_intercept()` or
+    /// `fit_intercept_and_normalize()` before calling `fit()`.
+    pub fn new() -> LinearRegression {
+        LinearRegression {
+            options: Options::None,
+        }
+    }
+
+    /// Configure the linear regression model to fit an intercept.
+    pub fn fit_intercept(mut self) -> Self {
```

Comment:
I don't really like the name, because `fit_intercept` sounds like a special function of `fit`, but is actually part of the builder. Perhaps it would be easier to have a `pub fn options(&mut self, options: Options)` or alternatively `with_intercept`.